NYPL/research-catalog
# SCC-3665: Server auth log in through cookie parsing and JWT verification

Resolves [SCC-3665](https://jira.nypl.org/browse/SCC-3665)

- Uses the current DFE `initializePatronTokenAuth` and `getPatronData` functions as reference:
  - `initializePatronTokenAuth`: https://github.com/NYPL/discovery-front-end/blob/development/src/server/routes/auth/index.js
  - `getPatronData`: https://github.com/NYPL/discovery-front-end/blob/development/src/server/routes/api/index.js
- Includes an additional function for `getPatronData` to get the 268 field for notification preference.
- Adds tests for these functions which didn't exist before.
- Right now, the patron data is only available on the homepage by calling `getPatronData` in the homepage `getServerSideProps`. Every page would need to make this function call again because I'm not setting the patron data to a global object -- and maybe we don't want to?

**Implementation notes**
In DFE, both `initializePatronTokenAuth` and `getPatronData` run on every request to every page. The patron decoded information from the cookie is set in Express' `request` object. This is then used in `getPatronData` to fetch `/patrons/${userId}` and the returned patron data is set in Redux.

For this implementation, a nextjs middleware is set with defined paths on where to run. For now it's the homepage and the search pages for testing. Instead of setting a property on Nextjs' `request` object with the patron data, it is being sent through Nextjs headers. These headers are then read in `getPatronData` and used in the `/patrons/${userId}` fetch call.

The headers are `"x-is-token-valid"` and `"x-user-id"`. From what I understand, these headers are "non-standard" and cannot be read by browsers. I was not able to view them in the browser or elsewhere but I do need someone to verify this. Otherwise, this implementation will have to change. I'm also not sure how secure sending these values through headers is.

**How to test**
There's some information in the readme. Make sure to test on `local.nypl.org:8080/research/research-catalog`.
- Log into qa or prod DFE.
- Start the app locally and open the homepage.
- The logged in patron data will be logged to the console.
- **This won't work on Vercel as expected due to the bounds of the `nyplIdentityPatron` cookie**

**Other updates**
- For now, removed next's `Link` component usage. It was throwing an error.
- Upgraded to Next13 for its middleware implementation.
- Upgraded React to v18 to be compatible with Next13 (using this with DS is ok but does output warnings).
- Adds the `jose` package for JWT verification. The middleware uses the edge runtime and not node (this is annoying), so the `jsonwebtoken` package could not be used.
- Removed `act` calls because those were making tests fail.
